ddtrace/internal/debug.py

```diff
@@ -0,0 +1,174 @@
+
+
+logger = get_logger(__name__)
+
+
+def ping_agent(api=None, hostname=None, port=None, uds_path=None):
+    # Attempt to query the agent, returns an api.Response
+    # or one of the following exceptions: httplib.HTTPException, OSError, IOError
+
+    if not api:
+        api = ddtrace.api.API(hostname=hostname, port=port, uds_path=uds_path,)
+
+    # We can't use api.send_traces([]) since it'll shortcut
+    # if traces is falsy.
+    p = ddtrace.payload.Payload(encoder=api._encoder)
+
+    # We can't use payload.add_trace([]) for the same reason
+    # as api.send_trace([]).
+    encoded = p.encoder.encode_trace([])
+    p.traces.append(encoded)
+    p.size += len(encoded)
+
+    try:
+        resp = api._flush(p)
+    except Exception as e:
+        resp = e
+
+    return resp
+
+
+def in_venv():
+    # Works with both venv and virtualenv
+    # https://stackoverflow.com/a/42580137
+    if "VIRTUAL_ENV" in os.environ:
+        return True
+
+    return hasattr(sys, "real_prefix") or (hasattr(sys, "base_prefix") and sys.base_prefix != sys.prefix)
+
+
+def tags_to_str(tags):
+    # Turn a dict of tags to a string "k1:v1,k2:v2,..."
+    return ",".join(["%s:%s" % (k, v) for k, v in tags.items()])
+
+
+def collect(tracer):
+    """Collect system and library information into a serializable dict.
+    """
+
+    # The tracer doesn't actually maintain a hostname/port, instead it stores
+    # it on the possibly None writer which actually stores it on an API object.
+    # Note that the tracer DOES have hostname and port attributes that it
+    # sets to the defaults and ignores afterwards.
+    if tracer.writer:
+        if isinstance(tracer.writer, writer.LogWriter):
+            agent_url = "AGENTLESS"
+            hostname = port = uds_path = None
+        else:
+            hostname = tracer.writer.api.hostname
+            port = tracer.writer.api.port
+            uds_path = tracer.writer.api.uds_path
+            https = tracer.writer.api.https
+
+            # If all specified, uds_path will take precedence
+            if uds_path:
+                agent_url = "uds://%s" % uds_path
+            else:
+                proto = "https" if https else "http"
+                agent_url = "%s://%s:%s" % (proto, hostname, port)
+    else:
+        # Else if we can't infer anything from the tracer, rely on the defaults.
+        hostname = tracer.hostname
+        port = tracer.port
+        agent_url = "http://%s:%s" % (hostname, port)
+
+    if (hostname and port) or uds_path:
+        resp = ping_agent(hostname=hostname, port=port, uds_path=uds_path)
+
+        if isinstance(resp, ddtrace.api.Response):
+            if resp.status == 200:
+                agent_error = None
+            else:
+                agent_error = "HTTP code %s, reason %s, message %s" % (resp.status, resp.reason, resp.msg)
+        else:
+            # There was an exception
+            agent_error = "Agent not reachable. Exception raised: %s" % str(resp)
```

Comment:
```suggestion
        try:
            resp = ping_agent(hostname=hostname, port=port, uds_path=uds_path)
        except Exception:
            agent_error = "Agent not reachable. Exception raised: %s" % str(resp)
        else:
            if resp.status == 200:
                agent_error = None
            else:
                agent_error = "HTTP code %s, reason %s, message %s" % (resp.status, resp.reason, resp.msg)
```

Reply:
`api._flush` actually returns exception objects itself so this would actually be 

```python
        try:
            resp = ping_agent(hostname=hostname, port=port, uds_path=uds_path)
        except Exception as e:
            agent_error = "Agent not reachable. Exception raised: %s" % str(e)
        else:
            if isinstance(resp, ddtrace.api.Response):
                if resp.status == 200:
                    agent_error = None
                else:
                    agent_error = "HTTP code %s, reason %s, message %s" % (resp.status, resp.reason, resp.msg)
            else:
                agent_error = "Agent not reachable. Exception raised: %s" % str(resp)
```

which is more complicated imo

Reply:
No because I also changed `api._flush`. So I stand by my suggestion. :)

Reply:
Yeah I saw both your suggested changes, but I was saying that [api._flush returns exceptions itself](https://github.com/DataDog/dd-trace-py/blob/206d6a8205f286158f484cb7ba23396727e59725/ddtrace/api.py#L246-L247). So we need to do the check.

Reply:
Oh I forgot about that weird design. Ok then, I guess we'll have to keep that 😫